src/node_api.cc

```diff
@@ -492,15 +492,51 @@ class TryCatch : public v8::TryCatch {
 
 //=== Function napi_callback wrapper =================================
 
-static const int kDataIndex = 0;
-static const int kEnvIndex = 1;
+static const int kFunctionIndex = 0;  // Used in CallbackBundle::cb[]
+static const int kGetterIndex = 0;    // Used in CallbackBundle::cb[]
+static const int kSetterIndex = 1;    // Used in CallbackBundle::cb[]
+static const int kCallbackCount = 2;  // Used in CallbackBundle::cb[]
+                                      // Max is "getter + setter" case
+
+static const int kCallbackBundleIndex = 0;  // The first and the only one
+static const int kInternalFieldCount = 1;
+
+// Use this data structure to reduce the number
+// of GetInternalField() calls to only 1 (was: 3).
+// This leads to better performance in runtime.
+// Ref: benchmark/misc/napi_function_call
+struct CallbackBundle {
+  ~CallbackBundle() {
+    if (handle.IsEmpty()) {
+      return;
+    }
+
+    handle.ClearWeak();
+    handle.Reset();
+  }
+
+  // Bind the lifecycle of `this` C++ object to a JavaScript object.
+  // We never delete a CallbackBundle C++ object directly.
+  void BindLifecycleTo(v8::Isolate* isolate, v8::Local<v8::Object> obj) {
+    handle.Reset(isolate, obj);
+    handle.SetWeak(this, WeakCallback, v8::WeakCallbackType::kParameter);
+  }
 
-static const int kFunctionIndex = 2;
-static const int kFunctionFieldCount = 3;
+  napi_env       env;      // Necessary to invoke C++ NAPI callback
+  void*          cb_data;  // The user provided callback data
+  napi_callback  cb[kCallbackCount];  // Max capacity is 2 (getter + setter)
+  v8::Persistent<v8::Object> handle;  // Die with this JavaScript object
```

Comment:
If you use a `node::Persistent` you can avoid having to `handle.ClearWeak()` and `handle.Reset()` in the destructor.